User interface for image uploads in editor tools

joaquinkunkel opened the pull request.

emilgoldsmith commented:
Hey @joaquinkunkel I actually wrote this in Slack a week ago when you pushed this branch; you haven't done a `git pull` on the master branch in a long time, so this branch is forked off a master branch that is 60 commits behind, which is probably the first time you cloned the repository. Please pull the latest master branch to your local clone, and rebase this branch onto that, I'll review your code then.

emilgoldsmith commented:
Hey @joaquinkunkel, for one if you could just fix the small linting error as that's what we have our CI for :).

And you rebased semi-correctly. The current code on the branch is probably correct now, but what you normally do when you rebase is what's called a force push that overwrites history, since `git rebase` changes history, you do this by doing `git push -f`, which should be used sparingly, and only on branches only you work on (for example see [here](https://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/) or many other guides/blog posts), but it's perfect for this case. 

So right now you can see that you are requesting to merge 2 copies of your commits and some of my old commits which is a bit messy :). You can probably use an interactive rebase (`git rebase -i master`) or `git reflog` to reset to before you merged in the old branch and do a force push then to fix it, feel free to look up some guides as well, this is kinda advanced work with git, but super useful when you understand it. If everything screws up completely also feel free to write and I can fix it.

emilgoldsmith commented:
@joaquinkunkel I just did you a bit of a favor and rebased your code again, fixing the weird history stuff and duplicates of commits that showed up, and also making it up to date with the big refactor I recently merged in (don't worry rebasing is an art, it's not easy at first). If you could just look over the PR one more time though and double check that all the changes are as you want them to be, that'd be great though (the rebase went pretty smoothly so I don't expect any problems, but you never know).

joaquinkunkel commented on src/components/admin/ImageController.jsx:
```diff
@@ -192,23 +200,31 @@ export default class ImageUploader extends BaseComponent {
+  }
+
+  changeImageName() {
+    const imgName = this.state.imgName;
```
Are there instances in which we would want to allow the user to change the file extension? In most if not all cases you would not want to change the extension of an image file directly on the website.

Before reading this, I changed the code so that it does the following:
(1) If the user types in the correct extension, it leaves it as is;
(2) If the user does not include an extension at all, it adds the original extension to the new filename; and
(3) If the user inputs the wrong extension, it changes it to the original one.

Of course, if this is not enough, I'll add what you suggested!

emilgoldsmith replied:
That sounds great, I think I agree the extension should never be changed on the website, I think we should do (3) a bit differently if you're just silently changing the extension though as that could be confusing, either give a notification with the snackbar or something else that the extension was changed back to the original or give them a warning that doesn't allow them to save.

joaquinkunkel commented on src/components/admin/ImageController.jsx:
```diff
@@ -56,7 +64,7 @@ export default class ImageUploader extends BaseComponent {
       const response = xhr.response;
       if (response.split(' ')[0] === 'success') {
-        const amazonURL = response.split(' ')[1];
+        const amazonURL = `https://s3.amazonaws.com/thegazelle/${response.split(' ')[1]}`;
```
My understanding at the time I did this was that the URL wasn't being entirely displayed and people had to copy and paste the whole `https://s3.amazonaws.com/thegazelle/` part (@zanemountcastle ?).
However, it used to show something other than `test_url` so I'm guessing that's changed now?

joaquinkunkel replied:
(I'll comment that part out for now.)

joaquinkunkel replied:
(I'll comment out that part for now.)

emilgoldsmith replied:
No it shows the full URL in production (when it actually uploads an image)

joaquinkunkel commented:
@zanemountcastle I changed the button style (now it looks more like a clickable button), and added logic to keep the extension even if the user doesn't input it.
About the two buttons on top of each other, I only see one instance of each button (checked by inspecting on Chrome). Are you sure there are two? Maybe I'm missing something.

emilgoldsmith commented:
@joaquinkunkel just tell me when you've addressed all the changes and would like another full review, okay?

joaquinkunkel commented on src/components/admin/ImageController.jsx:
```diff
@@ -272,84 +325,135 @@ export default class ImageUploader extends BaseComponent {
+          ${this.state.attemptedExt}, please convert it externally and re-upload.`}
+          autoHideDuration={6000}
+          onActionTouchTap={this.handleActionTouchTap}
+          onRequestClose={this.handleRequestClose}
```
Yup!

joaquinkunkel commented on src/components/admin/ImageController.jsx:
```diff
@@ -272,84 +325,135 @@ export default class ImageUploader extends BaseComponent {
+          <TextField
+            name="newName"
+            defaultValue={this.state.nameInput}
+            onChange={(event) => {
```
Messaging you on slack about this!

emilgoldsmith commented on src/components/admin/ImageController.jsx:
```diff
@@ -355,6 +361,8 @@ export default class ImageUploader extends BaseComponent {
       },
     };
 
+    let extensionErrorMsg = `To change the image's extension to . ${this.state.attemptedExt},`;
```
Isn't there now a space between the `.` and the extension? So it'll render like `. png` `. jpeg` instead of `.png` `.jpeg`?

emilgoldsmith commented on src/components/admin/ImageController.jsx:
```diff
@@ -438,21 +446,20 @@ export default class ImageUploader extends BaseComponent {
+          <form onSubmit={this.changeImageName}>
+            <TextField
+              name="newName"
+              fullWidth={this.state.fullWidthTextField}
```
We don't put constants in state, only mutables, so you actually just have to write `fullWidth` (which is React implicit for `fullWidth={true}`

joaquinkunkel commented:
Pushed those small changes!
